checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitAnnotatedTypeFactory.java

```diff
@@ -773,6 +777,38 @@ public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror
         }
     }
 
+    @Override
+    protected TypeAnnotator createTypeAnnotator() {
+        return new ListTypeAnnotator(super.createTypeAnnotator(), new NullnessTypeAnnotator(this));
+    }
+
+    /**
+     * This type annotator ensures that constructor return types are NONNULL, unless there is an
+     * explicit different annotation.
+     */
+    protected class NullnessTypeAnnotator extends TypeAnnotator {
+
+        /**
+         * Creates a new CommitmentTypeAnnotator.
+         *
+         * @param atypeFactory this factory
+         */
+        public NullnessTypeAnnotator(NullnessNoInitAnnotatedTypeFactory atypeFactory) {
+            super(atypeFactory);
+        }
+
+        @Override
+        public Void visitExecutable(AnnotatedExecutableType t, Void p) {
+            Void result = super.visitExecutable(t, p);
+            Element elem = t.getElement();
+            if (elem.getKind() == ElementKind.CONSTRUCTOR) {
+                AnnotatedDeclaredType returnType = (AnnotatedDeclaredType) t.getReturnType();
+                returnType.addMissingAnnotation(NONNULL);
```

Comment:
This should now only add NONNULL if there is no explicit annotation, even when the constructor comes from bytecode.
Can you open a follow-up issue that we should look whether this works correctly?
Have a file that has an explicit `@Nullable` on the constructor result and compile that file without the Nullness Checker. Then have a checked file that invokes that constructor. We should produce an error for that case.
You should be able to do this using a jtreg test case. See `checker/jtreg/nullness/constructor-initialization/NonDefaultConstructor.java`. (See note about `constructor-initialization` above. If there is no suitable directory, add something like `constructor-nullness`.